### ISSUE TITLE
Requires csp policy to allow "data:" because of the two background-image urls used for hover and active buttons.

### ISSUE THREAD

**Issue by 1ftskank** [sweetalert2/sweetalert2#877]
Sweetalert2 is great, but it uses some `background-image` data URLs to darken the button backgrounds on hover and when active:

```
.swal2-popup .swal2-actions:not(.swal2-loading) .swal2-styled:hover {
   background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR4nGNikAQAACIAHF/uBd8AAAAASUVORK5CYII="); }
.swal2-popup .swal2-actions:not(.swal2-loading) .swal2-styled:active {
   background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR4nGNiMAYAADwANpiOMBYAAAAASUVORK5CYII="); }
```

The upshot of this is that it requires the contentSecurityPolicy header to explicitly allow `data:` for the `img-src` policy. Is there any chance you would consider implementing a different way of darkening the images? I am worried about the implications of allowing `data:` for the `img-src` policy, for instance:

https://security.stackexchange.com/questions/94993/is-including-the-data-scheme-in-your-content-security-policy-safe/95011#95011

Maybe I am being too paranoid but wanted to raise the issue that using sweetalert2 requires a potential change to the content security policy header.

**Comment by limonte** [comment]
Thanks for reporting this important security detail @1ftskank!

IMO this is critical and should be fixed asap, it doesn't make sense to adjust server headers just because of some non-experienced front-end developer (me :man_facepalming:) decided to make things cool.

Any ideas how to make buttons' colors change dynamically for hover and active states are more than welcome!

**Comment by 1ftskank** [comment]
Great, thanks for being positive about this @limonte! I'm afraid I have no idea how it's normally done, but you may be able to get some funky js going:

https://stackoverflow.com/questions/5560248/programmatically-lighten-or-darken-a-hex-color-or-rgb-and-blend-colors

**Comment by 1ftskank** [comment]
Although you might run into other content security policy nightmares if you start changing inline styles programmatically, not sure.

**Comment by limonte** [comment]
JS tricks are tricky and, in practice, it requires A LOT of code in order to cover all cases from native `:hover` and `:active` states.

**We need pure CSS alternative.**

**Comment by 1ftskank** [comment]
Yeh fair enough, no idea then, sorry.

**Comment by limonte** [comment]
No problem, there are true CSS gurus around :) 

I'm just hanging around and helping to put together and release the awesomeness of the people from the community.

**Comment by samturrell** [comment]
We could just apply a `box-shadow` as a somewhat hacky but dynamic alternative e.g. `box-shadow: inset 0px 0px 0px 100px rgba(0,0,0,0.25);`

**Comment by acupajoe** [comment]
Thoughts?

```css
.swal2-popup button.swal2-styled {
  position: relative;
}

.swal2-popup button.swal2-styled:hover:after,
.swal2-popup button.swal2-styled:active:after {
  position: absolute;
  content: "";
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(0,0,0,0.15);
}
```

Demo: https://jsfiddle.net/kxxc7kxh/1/

**Comment by limonte** [comment]
Thanks @samturrell and @acupajoe for your ideas!

Unfortunately, both alternatives aren't good enough because they are changing not just the background color of buttons but the font color as well.

**Comment by limonte** [comment]
Help is still wanted. 

**Requirements:**

- pure CSS solution
- change only the background color of buttons, the visual font color should stay the same

**Comment by 1ftskank** [comment]
I've taken the example from @acupajoe and made it so only the background colour changes. It does involve making some changes the structure of your buttons (so doesn't come under the remit of "pure CSS solution", but once the button is in the correct format, it does only use CSS).

My solution uses a button as follows:

```
<button type="button" class="button">
  <div class='background'>
  </div>
  <div class="text">
    OK
  </div>
</button>
```

with the CSS looking like:

```
.button {
  position: relative;
  padding: 0;
}

.background {
  background-color: #3085d6;
  width: 120px;
  height: 60px;
}

.text {
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  margin: auto;
  position: absolute;
  color: white;
  font-size: 30pt;
  pointer-events: none;
}

.background:hover:after,
.background:active:after {
  position: absolute;
  content: "";
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: rgba(0,0,0,0.15);
}
```

I haven't got the vertically centering of the text down (I'm sure there is a way though), and it's obviously a bit of a change from the current structure of the DOM, but could be an idea:

https://jsfiddle.net/SamMiller/tc5rymuz/

**Comment by limonte** [comment]
What do you think guys about #884 ?

**Comment by birjolaxew** [comment]
[Here's a solution that doesn't require DOM changes.](https://jsfiddle.net/n9142zs6/)

**Comment by limonte** [comment]
Fixed in v7.9.2 

Thanks again @1ftskank for reporting this important issue!

**Comment by 1ftskank** [comment]
No worries, thanks for fixing! Awesome!
